src/librustc/infer/error_reporting/need_type_info.rs

```diff
@@ -136,6 +154,35 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             let expr = self.tcx.hir().expect_expr(body_id.hir_id);
             local_visitor.visit_expr(expr);
         }
+        let err_span = if let Some(pattern) = local_visitor.found_arg_pattern {
```

Comment:
Everything below here is moving from _compile vec of span labels and then create error_ to _create error as early as possible and operate on `DiagnosticBuilder` directly_.